### PR TITLE
fix(tidy3d): FXC-4563-use-2-d-intersections-for-adjoint-monitor-sizes-in-2-d-simulations

### PR DESCRIPTION
**Pull request by marcorudolphflex**
Previously, we just took the geometry bounds to construct adjoint monitors. This PR uses the plane intersection in 2D simulations.

<h2>Greptile Overview</h2>

<h3>Greptile Summary</h3>

This PR improves adjoint monitor sizing in 2D simulations by using planar geometry intersections instead of full bounding boxes. In 2D simulations (detected via `self.size.count(0.0) == 1`), the code now calls `geometry.intersections_plane()` to compute tight monitor bounds based on the actual geometry cross-section at the simulation plane.

Key changes:
- `simulation.py:4887` detects 2D simulations and passes plane info to structure monitor generation
- `structure.py:328-352` implements `_box_from_plane_intersection()` that computes monitor bounds from plane intersections with multi-level fallback (geometry → bounding box → full bounding box)
- Comprehensive tests validate behavior for spheres, meshes, cylinders, boxes, and polyslabs in both 2D and 3D cases
- Changelog entry added under 'Fixed' section

The implementation handles edge cases through a cascading fallback strategy and maintains backward compatibility for 3D simulations.

<h3>Confidence Score: 5/5</h3>

- This PR is safe to merge with comprehensive test coverage and proper fallback handling
- The changes are well-tested with multiple geometry types, include proper fallback logic for edge cases, maintain backward compatibility for 3D simulations, and fix an actual bug where adjoint monitors were oversized in 2D simulations. Code follows established patterns and includes appropriate documentation.
- No files require special attention

<h3>Important Files Changed</h3>

File Analysis

| Filename | Score | Overview |
|----------|-------|----------|
| CHANGELOG.md | 5/5 | Added changelog entry under 'Fixed' section documenting adjoint monitor cropping in 2D simulations |
| tests/test_components/autograd/test_adjoint_monitors.py | 5/5 | New comprehensive test file covering 2D and 3D adjoint monitor sizing with various geometries |
| tidy3d/components/simulation.py | 5/5 | Detects 2D simulations and passes plane info to structure monitor generation |
| tidy3d/components/structure.py | 4/5 | Implements plane intersection logic for 2D adjoint monitors, with fallback to bounding box |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Sim as Simulation
    participant Struct as Structure
    participant Geom as Geometry
    participant BoxCls as Box

    Note over Sim: User calls _make_adjoint_monitors
    Sim->>Sim: Check if 2D via size.count(0.0) == 1
    alt Is 2D simulation
        Sim->>Sim: sim_plane = self
    else Is 3D simulation
        Sim->>Sim: sim_plane = None
    end
    
    loop For each structure
        Sim->>Struct: _make_adjoint_monitors(freqs, index, field_keys, plane)
        
        alt plane is not None (2D case)
            Struct->>Struct: Call _box_from_plane_intersection
            Struct->>Geom: geometry.intersections_plane
            Geom-->>Struct: Return intersection shapes
            
            alt Intersections found
                Struct->>Struct: Compute union of bounds
            else No intersections
                Struct->>Geom: geom_box.intersections_plane
                alt Box intersections found
                    Struct->>Struct: Compute union of bounds
                else Still no intersections
                    Struct->>Struct: Use geom_box as fallback
                end
            end
            
            Struct->>BoxCls: Box.from_bounds(rmin, rmax)
            BoxCls-->>Struct: Return planar box
        else plane is None (3D case)
            Struct->>Struct: box = geom_box
        end
        
        Struct->>Struct: Create FieldMonitor with box
        Struct->>Struct: Create PermittivityMonitor with box
        Struct-->>Sim: Return monitors
    end
    
    Sim-->>Sim: Collect all monitors
```

---

> [!NOTE]
> Improves adjoint monitor sizing for 2D simulations by cropping to the simulation plane’s geometry intersection; 3D behavior unchanged.
> 
> - Simulation passes a planar context to structures in `_make_adjoint_monitors`; structures compute monitor `Box` via `geometry.intersections_plane(...)` with robust fallbacks
> - New tests validate 2D plane-cropped bounds (spheres, meshes incl. disjoint components) and confirm 3D still uses `bounding_box`
> - CHANGELOG updated; minor deterministic test data tweak (random seed)
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 119c999503ee615da22c73db3ea49da1d18a3220. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by marcorudolphflex** [comment]
@greptile

**Comment by marcorudolphflex** [comment]
@greptile

File: tests/utils.py [hunk 1]
```diff
@@ -1156,7 +1156,7 @@ def make_data(
     ) -> td.components.data.data_array.DataArray:
         """make a random DataArray out of supplied coordinates and data_type."""
         data_shape = [len(coords[k]) for k in data_array_type._dims]
-        np.random.seed(1)
+        np.random.seed(0)
```
Comment (marcorudolphflex) [inline tests/utils.py:1159]: that's a funny one. It caused a gradient being 2e-10 by chance in `test_multi_frequency_equivalence`

File: tidy3d/components/simulation.py [hunk 2]
```diff
@@ -4884,6 +4884,7 @@ def _make_adjoint_monitors(self, sim_fields_keys: list) -> tuple[list, list]:
             index_to_keys[index].append(fields)
 
         freqs = self._freqs_adjoint
+        sim_plane = self if self.size.count(0.0) == 1 else None
```
Comment (groberts-flex) [inline tidy3d/components/simulation.py:4887]: if we have adjoint monitors that are 3D, but the simulation is 2D, I believe the clipping will happen naturally in that the adjoint monitor will just have a 2D overlap with the simulation and so we will only get 2D data. I might be missing something though!

Comment (marcorudolphflex) [reply tidy3d/components/simulation.py:4887]: Yeah it does and this is fine most of the times. I target cases like here where the geometry bounds do not reflect the bounds within the sim plane: The non-centered Sphere has smaller bounds in the simulation plane. Same would apply for non-prism-like meshes.
Basically it is more about performance/efficiency as we may only shrink the monitor potentially. 

image

Comment (groberts-flex) [reply tidy3d/components/simulation.py:4887]: I see that makes sense!

**Review by yaugenst-flex — APPROVED** [review]
Sort of an edge case but nice to have, thanks @marcorudolphflex
